File: app/src/main/java/ru/samtakoy/core/screens/export_cards/BatchExportDialogFragment.java
package ru.samtakoy.core.screens.export_cards;

import android.app.Activity;
import android.app.Dialog;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AlertDialog;
import androidx.fragment.app.Fragment;

import javax.inject.Inject;
import javax.inject.Provider;

import moxy.MvpAppCompatDialogFragment;
import moxy.MvpPresenter;
import moxy.presenter.InjectPresenter;
import moxy.presenter.ProvidePresenter;
import ru.samtakoy.R;
import ru.samtakoy.core.MyApp;


public class BatchExportDialogFragment extends MvpAppCompatDialogFragment implements BatchExportDialogView{




    public static final String TAG = "BatchExportDialogFragment";

    private static final String ARG_EXPORT_TYPE = "ARG_EXPORT_TYPE";
    private static final String ARG_DIR_PATH = "ARG_DIR_PATH";

    private static final int EXPORT_TYPE_QPACKS = 1;
    private static final int EXPORT_TYPE_COURSES = 2;

    public static BatchExportDialogFragment newQPacksFragment(
            String dirPath, Fragment targetFragment, int targetReqCode
    ){
        BatchExportDialogFragment result = new BatchExportDialogFragment();
        Bundle args = new Bundle();
        args.putString(ARG_DIR_PATH, dirPath);
        args.putInt(ARG_EXPORT_TYPE, EXPORT_TYPE_QPACKS);

        result.setArguments(args);
        result.setTargetFragment(targetFragment, targetReqCode);
        return result;
    }

    // TODO пока экспорт только по Email, dirPath не используется
    public static BatchExportDialogFragment newCoursesFragment(
            String dirPath, Fragment targetFragment, int targetReqCode
    ){
        BatchExportDialogFragment result = new BatchExportDialogFragment();
        Bundle args = new Bundle();
        args.putString(ARG_DIR_PATH, dirPath);
        args.putInt(ARG_EXPORT_TYPE, EXPORT_TYPE_COURSES);

        result.setArguments(args);
        result.setTargetFragment(targetFragment, targetReqCode);
        return result;
    }


    @InjectPresenter
    MvpPresenter<BatchExportDialogView> mPresenter;
    @Inject
    Provider<BatchExportQPacksPresenter.Factory> mBatchExportQPacksPresenterProvider;
    @Inject
    Provider<BatchExportCoursesPresenter.Factory> mBatchExportCoursesPresenterProvider;

    @ProvidePresenter
    MvpPresenter<BatchExportDialogView> providePresenter() {
        Bundle args = getArguments();

        if (args.getInt(ARG_EXPORT_TYPE) == EXPORT_TYPE_QPACKS) {
            return mBatchExportQPacksPresenterProvider.get().create(args.getString(ARG_DIR_PATH));
        }
        return mBatchExportCoursesPresenterProvider.get().create(args.getString(ARG_DIR_PATH));
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {

        MyApp.getInstance().getAppComponent().inject(this);

        super.onCreate(savedInstanceState);
    }

    @NonNull
    @Override
    public Dialog onCreateDialog(@Nullable Bundle savedInstanceState) {

        View v = LayoutInflater.from(getContext()).inflate(R.layout.fragment_dialog_batch_export, null);
        return new AlertDialog.Builder(getContext())
                .setView(v)
                .setCancelable(false)
                .create();
    }



    @Override
    public void exitWithError(int errorTextId) {
        Toast.makeText(getContext(), errorTextId, Toast.LENGTH_SHORT).show();
        getTargetFragment().onActivityResult(getTargetRequestCode(), Activity.RESULT_CANCELED, null);
        dismiss();
    }

    @Override
    public void exitOk() {
        if(getTargetFragment() == null){
            return;
        }
        getTargetFragment().onActivityResult(getTargetRequestCode(), Activity.RESULT_OK, null);
        dismiss();
    }


}

File: app/src/main/java/ru/samtakoy/core/presentation/export_cards/BatchExportDialogFragment.kt
package ru.samtakoy.core.presentation.export_cards

import android.app.Activity
import android.app.Dialog
import android.os.Bundle
import android.view.LayoutInflater
import android.widget.Toast
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatDialogFragment
import androidx.core.os.bundleOf
import androidx.fragment.app.Fragment
import androidx.fragment.app.viewModels
import ru.samtakoy.R
import ru.samtakoy.core.app.di.Di
import ru.samtakoy.core.presentation.base.observe
import ru.samtakoy.core.presentation.base.viewmodel.AbstractViewModel
import ru.samtakoy.core.presentation.base.viewmodel.ViewModelOwner
import ru.samtakoy.core.presentation.export_cards.mv.BatchExportViewModel
import ru.samtakoy.core.presentation.export_cards.mv.BatchExportViewModelFactory
import ru.samtakoy.core.presentation.export_cards.mv.BatchExportViewModelImpl
import javax.inject.Inject

class BatchExportDialogFragment : AppCompatDialogFragment(), ViewModelOwner {

    @Inject
    internal lateinit var viewModelFactory: BatchExportViewModelFactory.Factory
    private val viewModel: BatchExportViewModelImpl by viewModels {
        viewModelFactory.factory(
            requireArguments().getParcelable<BatchExportType>(ARG_EXPORT_TYPE) as BatchExportType
        )
    }
    override fun getViewModel(): AbstractViewModel = viewModel

    override fun onCreate(savedInstanceState: Bundle?) {
        Di.appComponent.inject(this)

        super.onCreate(savedInstanceState)
    }

    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        val v = LayoutInflater.from(getContext()).inflate(R.layout.fragment_dialog_batch_export, null)
        return AlertDialog.Builder(requireContext())
            .setView(v)
            .setCancelable(false)
            .create()
    }

    override fun onObserveViewModel() {
        super.onObserveViewModel()
        viewModel.getViewActionsAsFlow().observe(viewLifecycleOwner, ::onAction)
    }

    private fun onAction(action: BatchExportViewModel.Action) {
        when (action) {
            BatchExportViewModel.Action.ExitOk -> exitOk()
            is BatchExportViewModel.Action.ExitWithError -> exitWithError(action.errorText)
        }
    }

    private fun exitWithError(errorText: String) {
        Toast.makeText(getContext(), errorText, Toast.LENGTH_SHORT).show()
        getTargetFragment()!!.onActivityResult(getTargetRequestCode(), Activity.RESULT_CANCELED, null)
        dismiss()
    }

    private fun exitOk() {
        if (getTargetFragment() == null) {
            return
        }
        getTargetFragment()!!.onActivityResult(getTargetRequestCode(), Activity.RESULT_OK, null)
        dismiss()
    }

    companion object {
        const val TAG: String = "BatchExportDialogFragment"

        private const val ARG_EXPORT_TYPE = "ARG_EXPORT_TYPE"

        fun newQPacksFragment(
            type: BatchExportType,
            targetFragment: Fragment,
            targetReqCode: Int
        ): BatchExportDialogFragment {
            val result = BatchExportDialogFragment()
            result.setArguments(
                bundleOf(ARG_EXPORT_TYPE to type)
            )
            result.setTargetFragment(targetFragment, targetReqCode)
            return result
        }

        fun newCoursesFragment(
            targetFragment: Fragment,
            targetReqCode: Int
        ): BatchExportDialogFragment {
            val result = BatchExportDialogFragment()
            result.setArguments(
                bundleOf(ARG_EXPORT_TYPE to BatchExportType.Courses)
            )
            result.setTargetFragment(targetFragment, targetReqCode)
            return result
        }
    }
}
